shelley/chain-and-ledger/executable-spec/test/Generator/Core.hs

```diff
@@ -0,0 +1,103 @@
+  ( findPayKeyPair
+  , genTxOut
+  , genUtxo0
+  , traceKeyPairs
+  , someKeyPairs
+  , pickStakeKey
+  , genInteger
+  , genNatural
+  , toAddr
+  , toCred
+  , unsafeMkUnitInterval)
+  where
+
+import           Cardano.Crypto.DSIGN (deriveVerKeyDSIGN, genKeyDSIGN)
+import           Crypto.Random (drgNewTest, withDRG)
+import           Data.Maybe (fromMaybe)
+import           Data.Word (Word64)
+import           Hedgehog (Gen)
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import           Numeric.Natural
+
+import           BaseTypes (UnitInterval, mkUnitInterval)
+import           Coin (Coin (..))
+import           Keys (pattern KeyPair, pattern SKey, pattern VKey, hashKey, vKey)
+import           LedgerState (genesisCoins)
+import           MockTypes (Addr, Credential, KeyPair, KeyPairs, SKey, TxOut, UTxO, VKey)
+import           Tx (pattern TxOut)
+import           TxData (pattern AddrBase, pattern KeyHashObj)
+
+mkKeyTuple :: (Word64, Word64, Word64, Word64, Word64) -> (VKey, SKey)
+mkKeyTuple seed = fst . withDRG (drgNewTest seed) $ do
+  sk <- genKeyDSIGN
+  return ( VKey $ deriveVerKeyDSIGN sk
+         , SKey sk)
+
+mkKeyPairs :: Word64 -> (KeyPair, KeyPair)
+mkKeyPairs n
+  = (mkKeyPair (2*n), mkKeyPair (2*n+1))
+  where
+    mkKeyPair n_ = (uncurry KeyPair) (mkKeyTuple (n_,n_,n_,n_,n_))
+
+-- | Constant list of KeyPairs intended to be used in the generators.
+traceKeyPairs :: KeyPairs
+traceKeyPairs = mkKeyPairs <$> [1 .. 200]
+
+-- | Select between _lower_ and _upper_ keys from 'traceKeyPairs'
+someKeyPairs :: Integer -> Integer -> Gen KeyPairs
+someKeyPairs lower upper = do
+  keys <- Gen.shuffle traceKeyPairs
+  n <- genInteger lower upper
+  return $ take (fromIntegral n) keys
+
+-- | Find first matching key pair for address. Returns the matching key pair
+-- where the first element of the pair matched the hash in 'addr'.
+findPayKeyPair :: Addr -> KeyPairs -> KeyPair
+findPayKeyPair (AddrBase (KeyHashObj addr) _) keyList =
+    fst $ head $ filter (\(pay, _) -> addr == hashKey (vKey pay)) keyList
+findPayKeyPair _ _ = error "currently no such keys should be generated"
+
+-- | Select one random verification staking key from list of pairs of KeyPair.
+pickStakeKey :: KeyPairs -> Gen VKey
+pickStakeKey keys = vKey . snd <$> Gen.element keys
+
+-- | Generator for a natural number between 'lower' and 'upper'.
+genNatural :: Natural -> Natural -> Gen Natural
+genNatural lower upper = Gen.integral $ Range.linear lower upper
+
+genInteger :: Integer -> Integer -> Gen Integer
+genInteger lower upper = Gen.integral $ Range.linear lower upper
+
+toAddr :: (KeyPair, KeyPair) -> Addr
+toAddr (payKey, stakeKey) = AddrBase (toCred payKey) (toCred stakeKey)
+
+toCred :: KeyPair -> MockTypes.Credential
+toCred k = KeyHashObj . hashKey $ vKey k
+
+-- | Generates a list of coins for the given 'Addr' and produced a 'TxOut' for each 'Addr'
+genTxOut :: [Addr] -> Gen [TxOut]
+genTxOut addrs = do
```

Comment:
:warning: This is a huge range! Like in the comment above, see what would be the smallest `TxOut`s list size that would allow us to get some good test coverage. If we have properties like no double spending maybe a list of size 10 might even suffice!

Reply:
this is a range of Coin value -  I've taken the 1000 down to 100 again

Reply:
I agree, that explains the big UTxOs I have seen in a first run.